src/windows/system.rs

```diff
@@ -236,107 +241,117 @@ impl SystemExt for System {
                     &mut cb_needed,
                 );
 
-                if ntstatus == STATUS_INFO_LENGTH_MISMATCH {
+                if ntstatus == STATUS_SUCCESS {
+                    break;
+                } else if ntstatus == STATUS_INFO_LENGTH_MISMATCH {
                     // GetNewBufferSize
                     if cb_needed == 0 {
-                        process_information.reserve(buffer_size);
                         buffer_size *= 2;
                         continue;
                     }
                     // allocating a few more kilo bytes just in case there are some new process
                     // kicked in since new call to NtQuerySystemInformation
                     buffer_size = (cb_needed + (1024 * 10)) as usize;
                     continue;
-                } else if ntstatus != STATUS_SUCCESS {
+                } else {
                     sysinfo_debug!(
                         "Couldn't get process infos: NtQuerySystemInformation returned {}",
                         ntstatus
                     );
-                    break;
+                    return;
                 }
+            }
+        }
 
-                process_information.set_len(buffer_size);
-
-                // Parse the data block to get process information
-                let mut process_ids = Vec::with_capacity(500);
-                let mut process_information_offset = 0;
-                loop {
-                    let p = process_information
-                        .as_ptr()
-                        .offset(process_information_offset)
-                        as *const SYSTEM_PROCESS_INFORMATION;
-                    let pi = &*p;
+        // If we reach this point NtQuerySystemInformation succeeded
+        // and the buffer contents are initialized
+        unsafe {
+            process_information.set_len(buffer_size);
+        }
 
-                    process_ids.push(Wrap(p));
+        // Parse the data block to get process information
+        let mut process_ids = Vec::with_capacity(500);
+        let mut process_information_offset = 0;
+        loop {
+            let p = unsafe {
+                process_information
```

Comment:
We could `assert!` that `process_information_offset` isn't too large here.

Reply:
Yes, could idea.

Reply:
I'll make a followup PR later with some more hardening.

Reply:
Fine by me!